[Mason] Vendingmachine Step3

myssun0325 opened the pull request:
자판기 3단계 구현했습니다.

(커밋 중 '3단계 리셋' : 이전까지 삽질한 작업 내용을 초기화하고, 이후 커밋부터 다시 작업하였습니다.)

myssun0325 committed:
Update Readme

myssun0325 committed:
테스트코드작성 : 자판기 객체 인스턴스화

myssun0325 committed:
코드구현: 자판기 구조체 생성

myssun0325 committed:
테스트코드작성 : 자판기의 넣은 돈이 처음에 0인지

myssun0325 committed:
코드구현: 자판기의 넣은 돈 속성 추가(0으로 초기화)

myssun0325 committed:
테스트코드작성: 자판기에 돈을 넣어 자판기의 투입된금액을 올린다

myssun0325 committed:
코드구현: 자판기에 돈을 넣는 메서드 작성

myssun0325 committed:
테스트코드작성: 특정 음료를 자판기에 추가

myssun0325 committed:
코드구현: 특정 음료를 자판기에 넣어 재고를 추가하는  메서드 작성

myssun0325 committed:
단위테스트 추가: 특정 음료를 자판기에 추가했을 때 재고의 갯수가 증가했는지 확인

myssun0325 committed:
단위테스트 추가: 다른 음료도 추가되는지 확인

myssun0325 committed:
단위테스트 추가: 여러 음료도 추가되는지 확인

myssun0325 committed:
테스트코드작성: 현재금액으로 구매가능한 음료수목록을 보여준다

myssun0325 committed:
Beverage배열을 객체화 -> BeverageList

myssun0325 committed:
beverageList의 객체가 비었는지 확인하는 연산프로퍼티 추가

myssun0325 committed:
리팩토링 VendingMachien의 하위 객체가 isEmpty를 수행하도록 변경

이에 따른 테스트코드 수정

myssun0325 committed:
투입된 금액으로 구매가능한 음료이름 출력 기능 추가

myssun0325 committed:
Rename method

myssun0325 committed:
딕셔너리 사용을 위한 Beverage객체의 Hashable프로토콜 준수

myssun0325 committed:
구매가능한 목록을 만드는 메서드 수정

myssun0325 committed:
테스트케이스 추가 : 구매가능한 음료가 여러개 있을 경우

myssun0325 committed:
테스트코드 변경으로 인한 불필요한 테스트삭제

myssun0325 committed:
구매가능한 음료수 목록을 리턴하는 메소드 리턴타입 변경

myssun0325 committed:
테스트코드 수정: 구매가능함 음료수 목록 확인을 위한 목록의 갯수 체크

myssun0325 committed:
Rename : purchasable -> buyable로 변경

myssun0325 committed:
테스트코드작성: 음료를 추가하고 돈을 넣은 뒤 음료를 구매한다.

myssun0325 committed:
불필요한 프로토콜 채택 삭제

myssun0325 committed:
dictionary타입 추상화를 위한 BeverageStock 객체 추가

myssun0325 committed:
BeverageStock, BeverageList객체 타입 활용하도록 수정

myssun0325 committed:
3단계 첫번째 삽질 Reset (다시 처음부터 연습)

myssun0325 committed:
VendingMachine객체 생성

myssun0325 committed:
실패테스트 작성

자판기에 원하는 금액을 넣고 자판기의 잔액을 통해 메소드 결과 확인

myssun0325 committed:
테스트통과 코드 작성

insertMoney 메소드, balance(잔액) 프로퍼티 구현

myssun0325 committed:
실패테스트 작성

자판기의 잔액을 확인.

myssun0325 committed:
테스트통과 코드 작성

잔액을 확인하는 메서드

myssun0325 committed:
리팩토링 테스트코드

checkBalance메소드가 테스트를 통과했으므로 insertMoney메소드의 결과를
checkBalance메소드를 통해 확인

myssun0325 committed:
속성에 private 접근제어 추가

myssun0325 committed:
실패테스트 작성

자판기가 음료수재고 속성을 갖고 있고, 처음에 비어있어야한다.

myssun0325 committed:
음료수의 카테고리 이름을 추상화 하기 위한 BeverageName 타입 정의 추가

myssun0325 committed:
Beverage객체가 NSObject를 상속함

Equatable과 Hashable과 같은 프로토콜을 사용하기 위해 Beverage객체가
CustomStringConvertible의 프로토콜을 채택하는 것이 아닌 NSObject를
상속함

myssun0325 committed:
테스트통과 코드 작성

Beverage 배열을 BeverageList로 추상화
자판기의 재고 딕셔너리를 BeverageStock으로 추상화
자판기의 처음 재고 상태는 BeverageStock 초기화 상태와 같음

myssun0325 committed:
실패테스트 작성

자판기에 음료수를 추가하면 자판기의 재고가 변한다.

myssun0325 committed:
테스트통과 코드 작성

특정 상품 인스턴스를 넘겨서 재고를 추가하는 메서드
VendingMachine의 add(beverage:) 메서드 구현

myssun0325 committed:
테스트케이스 추가

여러 음료 추가 확인

myssun0325 committed:
BeverageName의 case 이름 변경

myssun0325 committed:
테스트코드 버그 수정

myssun0325 committed:
실패테스트 작성

현재 금액으로 구매가능한 음료수 목록을 리턴받는 메소드

myssun0325 committed:
-------------------3단계 리셋-----------------

3단계 세번째 삽질을 위해 두번째 삽질 다시 리셋
This reverts commit 9c007013a5ca5a0d4278c29208da27386fd8412e.

myssun0325 committed:
Beverage가 NSObject를 상속받음

Beverage의 descriptioin구현이 CustomStringConvertible에서 NSObject상속을
활용한 방법으로 변경

myssun0325 committed:
VendingMachine구조체 생성

myssun0325 committed:
자판기 금액을 원하는 금액만큼 올리는 메소드 구현

잔액 balance 속성 추가

myssun0325 committed:
자판기 잔액을 확인하는 메소드 구현

myssun0325 committed:
재고(stock)속성 추가

재고속성을 위한 nested type의 ProductType enum 정의

myssun0325 committed:
재고 속성의 타입을 Stock타입으로 객체화

myssun0325 committed:
ProductType enum 이동

myssun0325 committed:
Stock -> StockManager로 변경

재고를 StockManager가 관리

myssun0325 committed:
테스트를 위해 각 Beverage에 convenience initializer 추가

myssun0325 committed:
특정 상품 인스턴스를 넘겨서 재고를 추가하는 메소드 구현

[Beverage]를 객체화하기 위해 Products로 표현

myssun0325 committed:
테스트 파일 추가

myssun0325 committed:
현재 금액으로 구매가능한 음료수 목록을 리턴하는 메소드 구현

- ProductType에 타입별 default price 속성 추가
- 음료수 객체 생성 시 default price와 생성날짜로 초기화하는 convenience
init 추가
- 구현 메소드 테스트케이스 추가

myssun0325 committed:
음료수를 구매하는 메소드 구현

- 음료수가 비어있을 경우 nil을 반환
- 음료수가 있을 경우 금액이 차감되고 구매한 음료수타입의 객체 반환

myssun0325 committed:
전체 상품 재고를 (사전으로 표현하는) 종류별로 리턴하는 메소드

myssun0325 committed:
유통기한이 지난 재고만 리턴하는 메소드 구현

- Beverage에 유통기한을 체크하는 메서드와 Equatable 구현
- ProductType의 유통기한을 반환하는 computed property 구현 및 비교를
위한 프로토콜 채택
- stock과 관련된 객체들을 구조체에서 클래스로 변경
- 클래스로 변경 후 NSObject를 상속

myssun0325 committed:
음료수를 구매하는 메소드를 수정

구매하고자하는 음료수를 다 구매하면
재고목록에서 삭제 및 없을 경우 nil 반환

myssun0325 committed:
isHot 속성 추가

myssun0325 committed:
따뜻한 음료만 리턴하는 메소드 구현(리팩토링 포함)

- 기존의 특정 조건(ex:유통기한이 지난)을 만족하는 음료를 반환받는
메소드가 계속 겹쳐서 조건을 클로저로 넘겨주면 그 조건에 맞는 음료를
반환받는 메소드로 수정함

myssun0325 committed:
StockManager의 클로저 구문 수정

myssun0325 committed:
구매 상품 이력을 저장하고 리턴하는 메소드 구현

myssun0325 committed:
자판기 struct -> class 변경

myssun0325 committed:
자판기 메소드 및 타입 변경에 따른 테스트코드 수정

myssun0325 committed:
음료 종류의 한글 표기 description 추가

myssun0325 committed:
StockManager에 특정 음료종류의 재고를 반환하는  메소드 추가

myssun0325 committed:
VendingMachine 메소드 수정 및 추가

- 재고가 모자를 경우와 잔액이 모자를 경우의 에러 정의
- 특정 타입의 재고를 반환하는 메소드 추가
- 음료수 구매 시 재고와 잔액을 확인하여 에러를 반환하고 구매 성공시
구매한 음료수를 반환하도록 메소드 수정

myssun0325 committed:
사용자 입력을 받기 위한 InputView 구조체 추가

메뉴를 선택하고 돈을 넣을 수 있는 메소드 정의

myssun0325 committed:
입력과 출력 형식을 맞추기 위한 Formatter 구조체 추가

myssun0325 committed:
출력을 위한 OutputView 구조체 추가

외부에서 VendingMachine을 받아서 사용하도록 구현

myssun0325 committed:
시나리오 구현

myssun0325 committed:
리팩토링 : 타입 의존 -> 프로토콜 의존으로 수정

- OutputView가 AvailableVendingMachine 프로토콜을 의존하도록 구현
- VendingMachine이 StockManagable 프로토콜을 의존하도록 구현
- main에서 사용자가 구매 불가능한 음료 채택 시 오류처리

godrm commented on VendingMachine/VendingMachine/Beverage.swift:
```diff
@@ -8,29 +8,47 @@
         self.price = price
         self.name = name
         self.manufacturedDate = manufacturedDate.customDateFormat
+        self.isHot = isHot
```
isHot 같은 속성이 모든 음료수에 다 들어가는게 적당할까요?
커피처럼 따뜻하게 먹는 음료수만 들어가면 될까요? 
그리고 왜 이 속성은 private(set) 인가요?

myssun0325 replied:
질문있습니다.
isHot을 모든 속성이 아닌 특정 음료에만 적용하도록 수정중입니다.
음료가 isHot 속성을 가지고 있을 경우 속성자체가 Bool이기 때문에
메서드를 통해서 이 값을 그대로 취해오는 경우와 차이가 없는거 같아서 private(set)을 사용했습니다.
이 경우에도 비록 메서드에서 `return self.isHot`을 해주더라도 메서드를 통해 가져오는게 나은건가요?~
(이 부분의 속성은 외부에서 이 값의 true인지  false인지를 판단하여 제거할 때 사용됩니다.)

godrm replied:
좀 더 근본적인 의미에서 설계에 대해 고민해보세요.
해당 음료가 뜨거운지 차가운지 외부에서 값을 받아서 비교하는 행위가 필요한가 입니다.

myssun0325 replied:
방금 주신 코멘트 부분에 대해서 이렇게 생각해보았습니다.

속성으로 온도를 가지고 있고 이를 사용하고자 하는 메서드의 인자로 외부에서 확인하고자 하는 온도를 받은다음
이 온도보다 높으면 Hot이고 아니면 false가 되는 구조를 떠올려보았습니다.

godrm commented on VendingMachine/VendingMachine/Formatter.swift:
```diff
@@ -0,0 +1,21 @@
+
+import Foundation
+
+struct Formatter {
```
Formatter 도 구체적으로 어떤 표현을 위한 포맷을 담당하는지 명시하면 좋겠습니다.

godrm commented on VendingMachine/VendingMachine/Milk/Milk.swift:
```diff
@@ -13,9 +13,9 @@ class Milk: Beverage {
     private let expiredDate: Date
     private let expiredPeriod: TimeInterval = 864000 // 10 days
```
상수 값을 주석으로 설명하기 보다 단위를 표현할 수 있는 방법은 없었을까요?

godrm commented on VendingMachine/VendingMachine/OutputView.swift:
```diff
@@ -0,0 +1,55 @@
+    }
+    
+    func printSoldBeverage(_ productType: ProductType) throws {
+        let beverage = try self.vendingMachine.buy(productType)
```
이 부분은 역할 분담이 맞는걸까요? 
실제로 OutputView가 구매를 처리하고 있네요. 구매하고 결과만 출력하도록 로직과 분리하세요.

godrm commented on VendingMachine/VendingMachine/ProductType.swift:
```diff
@@ -0,0 +1,80 @@
+    case Cantata
+    case Georgia
+    case StrawberryMilk
+    case ChocoMilk
```
Beverage에서 상속받아서 만든 타입이 있는데 여기 타입들을 또 만들었네요?
이렇게 되면 음료가 바뀌면 양쪽다 바꿔줘야 하는 셈입니다. 
이 타입이 꼭 필요한가요? 추가적인 타입은 없는게 더 좋은 설계입니다.

godrm commented on VendingMachine/VendingMachine/VendingMachine.swift:
```diff
@@ -0,0 +1,94 @@
+    }
+    
+    func readBalance() -> String {
+        return "\(self.balance)원"
```
여기처럼 값과 `단위`는 분리해서 처리하는 게 좋습니다.
로직을 구현하는 VendingMachine은 값만 넘겨주고, 출력하는 단위 등은 OutputView만 처리하는게 좋습니다.

godrm reviewed (commented):
ProductType 에 대해 다시 한 번 고민해보세요.
피드백을 확인해보세요.

myssun0325 committed:
피드백반영: 음료의 isHot 속성 수정

커피객체만 뜨거움을 나타내는 속성을 갖도록 표현
isHot이 아닌 온도를 속성으로 갖고 외부에서 이 값에 대해서
뜨거운지 안뜨거운지는 값을 주어 판단하여 반환하도록 설정

myssun0325 committed:
피드백반영: 주석없이 단위를 표현하기

myssun0325 committed:
피드백반영: VendingMachine의 잔액처리를 값과 단위를 분리

VendingMachine은 값만 넘겨주고 단위는 OutputView에서 정하도록

myssun0325 committed:
피드백 반영

- ObjectIdentifier 적용
- isHot속성을 커피에만 적용
- 구매가능한 음료의 반환값으로 Formatter를 사용하지 않기 위해 Beverage들을 갖고 있는 Product의 배열을 반환하도록 변경
- 유통기한의 상수 값 표현 변경
- outputview에서 구매하지않도록 변경 -> 구매한 결과만 출력
- OutputView에서 값과 단위를 따로 처리

myssun0325 committed:
Formatter삭제: 반환타입 변경으로 불필요해진 객체

myssun0325 commented:
피드백 반영 하였습니다.

- key값으로 사용되던 ProductType을 제거하고 ObjectIdentifier 적용
- isHot속성을 커피에만 적용 -> 음료가 온도를 갖고 있고 외부에서 입력받은 온도보다 높은지를 판단해주는 isHot메소드 로 변경
- 구매가능한 음료의 반환값으로 Formatter를 사용하지 않기 위해 Beverage들을 갖고 있는 Product의 배열을 반환하도록 변경 -> Product가 음료들의 정보를 갖고 있음.
- 유통기한의 상수 값 표현 변경 -> 유통기한이라는 이름 사용 및 값 표현: day * 하루를 초로 나타내도록
- outputview에서 구매하지않도록 변경 -> 구매한 결과만 출력하도록 변경
- OutputView에서 값과 단위를 따로 처리 -> 값은 VendingMachine이 단위는 OutputView에서

godrm commented on VendingMachine/VendingMachine/Milk/Milk.swift:
```diff
@@ -11,10 +11,10 @@ import Foundation
     
     private let expiredDate: Date
-    private let expiredPeriod: TimeInterval = 864000 // 10 days
+    private let tenDaysPeriod: TimeInterval = 86400 * 10
```
이런 값은 static 으로 선언해도 좋습니다. 참고하세요.

godrm commented:
다음 단계에 도전해봅시다. 큰 도전꺼리가 기다릴겁니다 :)
